FaunaDB.Client/Types/Encoder.cs

```diff
@@ -0,0 +1,278 @@
+                throw new InvalidOperationException($"Self referencing loop detected for object `{obj}`");
+
+            try
+            {
+                stack.Push(obj);
+
+                return EncodeIntern(obj);
+            }
+            finally
+            {
+                stack.Pop();
+            }
+        }
+
+        Value EncodeIntern(object obj)
+        {
+            var type = obj.GetType();
+
+            switch (Type.GetTypeCode(type))
+            {
+                case TypeCode.String:
+                    return StringV.Of((string)obj);
+
+                case TypeCode.Boolean:
+                    return BooleanV.Of((bool)obj);
+
+                case TypeCode.DateTime:
+                    return WrapDateTime((DateTime)obj);
+
+                case TypeCode.SByte:
+                case TypeCode.Int16:
+                case TypeCode.Int32:
+                case TypeCode.Int64:
+                    return LongV.Of(Convert.ToInt64(obj));
+
+                case TypeCode.Char:
+                case TypeCode.Byte:
+                case TypeCode.UInt16:
+                case TypeCode.UInt32:
+                case TypeCode.UInt64:
+                    return LongV.Of(Convert.ToInt64(obj));
+
+                case TypeCode.Single:
+                case TypeCode.Double:
+                case TypeCode.Decimal:
+                    return DoubleV.Of(Convert.ToDouble(obj));
+
+                case TypeCode.Empty:
+                    return NullV.Instance;
+
+                case TypeCode.Object:
+                    if (typeof(byte[]).IsInstanceOfType(obj))
+                        return new BytesV((byte[])obj);
+
+                    if (typeof(IDictionary).IsInstanceOfType(obj))
+                        return WrapDictionary((IDictionary)obj);
+
+                    if (typeof(IEnumerable).IsInstanceOfType(obj))
+                        return WrapEnumerable((IEnumerable)obj);
+
+                    return WrapObj(obj);
+            }
+
+            throw new InvalidOperationException($"Unsupported type {type} at `{obj}`");
+        }
+
+        Value WrapDateTime(DateTime date)
+        {
+            if (date.Ticks % (24 * 60 * 60 * 10000) > 0)
+                return new TimeV(date);
+
+            return new DateV(date);
+        }
+
+        Value WrapEnumerable(IEnumerable array)
+        {
+            var ret = new List<Value>();
+
+            foreach (var value in array)
+                ret.Add(Encode(value));
+
+            return new ArrayV(ret);
+        }
+
+        Value WrapDictionary(IDictionary dic)
+        {
+            var ret = new Dictionary<string, Value>();
+
+            foreach (DictionaryEntry entry in dic)
+                ret.Add(entry.Key.ToString(), Encode(entry.Value));
+
+            return new ObjectV(ret);
+        }
+
+        Value WrapObj(object obj)
+        {
+            var type = obj.GetType();
+
+            Converter converter;
+
+            lock (converters)
```

Comment:
since this is Write-seldom, read-often, should we use a https://msdn.microsoft.com/en-us/library/system.threading.readerwriterlockslim(v=vs.110).aspx here?